examples/thermostat/thermostat-common/src/atomic-write-manager.cpp

```diff
@@ -0,0 +1,535 @@
+    auto attributeIdsIter = commandData.attributeRequests.begin();
+    while (attributeIdsIter.Next())
+    {
+        auto & attributeId = attributeIdsIter.GetValue();
+        attributeRequestCount++;
+        switch (attributeId)
+        {
+        case Presets::Id:
+            if (requestedPresets) // Double-requesting an attribute is invalid
+            {
+                return imcode::InvalidCommand;
+            }
+            requestedPresets = true;
+            break;
+        case Schedules::Id:
+            if (requestedSchedules) // Double-requesting an attribute is invalid
+            {
+                return imcode::InvalidCommand;
+            }
+            requestedSchedules = true;
+            break;
+        default:
+            // TODO: If this is a valid attribute on thermostat, but just isn't atomic, we shouldn't return an error code here
+            return imcode::InvalidCommand;
+        }
+    }
+    if (attributeIdsIter.GetStatus() != CHIP_NO_ERROR)
+    {
+        return imcode::InvalidCommand;
+    }
+    if (requestedPresets && requestedSchedules)
+    {
+        return imcode::Success;
+    }
+    return imcode::InvalidCommand;
+}
+
+ScopedNodeId GetSourceScopedNodeId(CommandHandler * commandObj)
+{
+    ScopedNodeId sourceNodeId = ScopedNodeId();
+    auto sessionHandle        = commandObj->GetExchangeContext()->GetSessionHandle();
+
+    if (sessionHandle->IsSecureSession())
+    {
+        sourceNodeId = sessionHandle->AsSecureSession()->GetPeer();
+    }
+    else if (sessionHandle->IsGroupSession())
+    {
+        sourceNodeId = sessionHandle->AsIncomingGroupSession()->GetPeer();
+    }
+    return sourceNodeId;
+}
+
+/**
+ * @brief Schedules a timer for the given timeout in milliseconds.
+ *
+ * @param[in] endpoint The endpoint to use.
+ * @param[in] timeoutMilliseconds The timeout in milliseconds.
+ */
+void ThermostatAtomicWriteManager::ScheduleTimer(EndpointId endpoint, System::Clock::Milliseconds16 timeout)
+{
+    DeviceLayer::SystemLayer().StartTimer(timeout, TimerExpiredCallback,
+                                          reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+/**
+ * @brief Clears the currently scheduled timer.
+ *
+ * @param[in] endpoint The endpoint to use.
+ */
+void ThermostatAtomicWriteManager::ClearTimer(EndpointId endpoint)
+{
+    DeviceLayer::SystemLayer().CancelTimer(TimerExpiredCallback, reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+void ThermostatAtomicWriteManager::OnTimerExpired(EndpointId endpoint)
+{
+    for (size_t i = 0; i < ArraySize(mAtomicWriteSessions); ++i)
+    {
+
+        auto atomicWriteState = mAtomicWriteSessions[i];
+        if (atomicWriteState.state == kAtomicWriteState_Open && atomicWriteState.endpointId == endpoint)
+        {
+            if (mDelegate != nullptr)
+            {
+                mDelegate->OnRollbackWrite(endpoint, Presets::Id);
```

Comment:
Can we have the delegate api just take endpoint as an argument and the delegate can rollback both if applicable. Calling this twice just looks weird mDelegate->OnRollbackWrite(endpoint)